pennylane/vqe/vqe.py

```diff
@@ -161,66 +160,55 @@ def __init__(
         """Bool: Records if multiple devices are input"""
 
-        if np.isclose(qml.math.toarray(qml.math.count_nonzero(coeffs)), 0):
-            self.cost_fn = lambda *args, **kwargs: np.array(0)
-            return
-
-        self._optimize = optimize
-
-        self.qnodes = qml.map(
-            ansatz, observables, device, interface=interface, diff_method=diff_method, **kwargs
-        )
-
-        if self._optimize:
+        self.qnodes = []
+        """The QNodes to be evaluated."""
 
+        if optimize:
             if self._multiple_devices:
                 raise ValueError("Using multiple devices is not supported when optimize=True")
 
-            obs_groupings, coeffs_groupings = qml.pauli.group_observables(observables, coeffs)
-            d = device[0] if self._multiple_devices else device
-            w = d.wires.tolist()
+            hamiltonian.compute_grouping()
 
             @qml.qnode(device, interface=interface, diff_method=diff_method, **kwargs)
-            def circuit(*qnode_args, obs, **qnode_kwargs):
-                """Converting ansatz into a full circuit including measurements"""
-                ansatz(*qnode_args, wires=w, **qnode_kwargs)
-                return [qml.expval(o) for o in obs]
-
-            def cost_fn(*qnode_args, **qnode_kwargs):
-                """Combine results from grouped QNode executions with grouped coefficients"""
-                if device.shot_vector:
-                    shots_batch = sum(i[1] for i in device.shot_vector)
-
-                    total = [0] * shots_batch
-
-                    for o, c in zip(obs_groupings, coeffs_groupings):
-                        res = circuit(*qnode_args, obs=o, **qnode_kwargs)
-
-                        for i, batch_res in enumerate(res):
-                            total[i] += sum(batch_res * c)
-                else:
-                    total = 0
-                    for o, c in zip(obs_groupings, coeffs_groupings):
-                        res = circuit(*qnode_args, obs=o, **qnode_kwargs)
-                        total += sum(r * c_ for r, c_ in zip(res, c))
-                return total
+            def circuit(params, **circuit_kwargs):
+                ansatz(params, wires=device.wires, **circuit_kwargs)
+                return qml.expval(hamiltonian)
 
-            self.cost_fn = cost_fn
+            self.qnodes.append(circuit)
+            self.cost_fn = self.qnodes[0]
 
         else:
-            self.cost_fn = qml.dot(coeffs, self.qnodes)
+            coeffs, observables = hamiltonian.terms()
+            if not self._multiple_devices:
+                device = [device] * len(coeffs)
+
+            for obs, dev in zip(observables, device):
+                wires = dev.wires
+
+                @qml.qnode(
+                    dev,  # pylint: disable=cell-var-from-loop
+                    interface=interface,
+                    diff_method=diff_method,
+                    **kwargs,
+                )
```

Comment:
```suggestion
            if not self._multiple_devices:

            for obs in observables:
                wires = device.wires

                @qml.qnode(
                    device,
                    interface=interface,
                    diff_method=diff_method,
                    **kwargs,
                )
```

Reply:
What is this suggestion for?

Reply:
Since the same device is being used for all the qnodes in the loop, I don't see the point of creating a list with the same device to iterate over.

Reply:
It's mainly to avoid copying the for loop twice

Reply:
Otherwise we'd have a lot of copied lines

Reply:
I'm not sure I understand what you mean. The loop would remain the same, but it would just iterate over the observables rather than both the observables and the device, as the device is the same for all qnodes. Additionally, since all qnodes use the same device, you could create the qnode outside the for loop and pass in the observable as a parameter to the qnode.

Reply:
That's why we check if `device` is a list or not, and convert it to a list if needed.

Reply:
I'm mainly copying the code from `qml.map` (because `map` is now deprecated).